### ISSUE TITLE
[BUG] npm i -g @swc/cli removed my git repository

### ISSUE THREAD

**Issue by kdy1** [npm/cli#3457]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Deletes git repository.
(Removed `.git` and tracked files)

### Expected Behavior

Install without error

### Steps To Reproduce

1. Clone https://github.com/kdy1/swc.git
2. Checkout `resolver` branch
3. npm i
4. npm run build:dev
5. npm link
6. npm i -g @swc/cli

### Environment

- OS: macOS Big Sur
- Node: 

```
➜  swc-bugs node -v
v14.16.1
```

Note: I installed latest version of node js because of stack trace issue on github actions.

```
➜  swc-bugs brew info node
node: stable 16.3.0 (bottled), HEAD
Platform built on V8 to build network applications
https://nodejs.org/
/usr/local/Cellar/node/14.4.0 (107 files, 650KB)
  Built from source
/usr/local/Cellar/node/15.12.0 (3,268 files, 55.5MB)
  Poured from bottle on 2021-03-31 at 15:11:58
/usr/local/Cellar/node/16.2.0 (2,784 files, 49MB) *
  Poured from bottle on 2021-05-30 at 20:06:55
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/node.rb
License: MIT
==> Dependencies
Build: pkg-config ✔, python@3.9 ✔
Required: brotli ✔, c-ares ✔, icu4c ✔, libuv ✔, nghttp2 ✔, openssl@1.1 ✔
==> Options
--HEAD
	Install HEAD version
==> Analytics
install: 323,426 (30 days), 1,082,930 (90 days), 4,400,808 (365 days)
install-on-request: 254,200 (30 days), 847,506 (90 days), 3,304,651 (365 days)
build-error: 0 (30 days)
```

- npm:
```
npm@7.14.0 /Users/kdy1/.nvm/versions/node/v14.16.1/lib/node_modules/npm
```

**Comment by ljharb** [comment]
Duplicate of #2104?

**Comment by kdy1** [comment]
@ljharb `npm link` was successful. Before getting this issue, my folder was removed when I **canceled** `npm i -g @swc/cli`

I get this issue even though `@swc/cli` installation was successful.

**Comment by isaacs** [comment]
I can reproduce this.

Simplest reproduction I've whittled it down to so far:

```bash
#!/usr/bin/env bash
rm -rf prefix swc
mkdir -p prefix/lib
mkdir swc

# just a dummy file
cat >swc/package.json <<PJ
{
  "name": "@swc/core",
  "version": "1.2.23-alpha.1"
}
PJ

cat >swc/donotdelete <<DATA
please do not delete this file
DATA

args=(--quiet --audit=false --ignoreScripts --global)
arborist reify prefix/lib --add=file:swc "${args[@]}"

arborist reify prefix/lib "${args[@]}" --update # '--add=@swc/cli' also works

if ! [ -d swc/donotdelete ]; then
  echo 'DAMAGED'
  exit 1
else
  echo 'PASS'
fi
```

This seems related to a few different subtle issues combining.  I _can't_ seem to get it to reproduce in a simpler tree, so there's definitely something very interesting here relating (i think?) to a peer dep of both a top-level global package and one of its dependencies, which is not satisfied by the thing being linked, and where the peer dep itself has dependencies as well.

**Comment by isaacs** [comment]
Ok, I have a fix for this.  It'll be in the next npm release.  Thanks @nlf for pairing through it with me.

**Comment by isaacs** [comment]
This is fixed in npm 7.21.0
